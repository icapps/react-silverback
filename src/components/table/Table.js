import React from 'react';
import PropTypes from 'prop-types';
import format from 'date-fns/format';
import isDate from 'date-fns/is_date';
import { strings } from '../../utils';
import { Modal } from '../index';
import { identifiers } from '../../constants';
import './table.css';

const checkIcon = require('../../assets/images/check.svg');
const crossIcon = require('../../assets/images/cross.svg');
const deleteIcon = require('../../assets/images/delete.svg');

const SORT_ASC = 'asc';
const SORT_DESC = 'desc';

class Table extends React.Component {
  sort = (sortField, isSortable) => {
    if (isSortable) {
      const sortOrder = (this.props.sortField === sortField) ? (this.props.sortOrder === SORT_ASC ? SORT_DESC : SORT_ASC) : SORT_ASC;
      this.props.handleSort(sortField, sortOrder);
    }
  }

  renderData = data => {
    switch (typeof data) {
      case 'object': return isDate(data) ? format(data, this.props.dateFormat) : data.toString();
      case 'string': return data.length < this.props.maxTextLength ? data : `${data.substring(0, this.props.maxTextLength)}...`;
      case 'boolean': return <img src={data ? checkIcon : crossIcon} alt={`${data}`} />;
      default: return data;
    }
  }

  selectData = (object, path) => {
    return this.renderData(path.split('.').reduce((obj, prop) => {
      return obj[prop];
    }, object
    ));
  }

  render() {
    const { props } = this;
    return (
      <div className="table-responsive">
        <table className="table table-hover">
          <thead>
            <tr>
              {props.keys.map(key => (
                <th scope="col" key={key.id}>
                  <span className={`key ${key.isSortable ? 'sortable-key' : ''} ${props.sortField === key.sorter ? 'active-key' : ''}`} onClick={() => this.sort(key.sorter, key.isSortable)}>
                    {key.value}
                    {key.isSortable && <span className={`sort ${props.sortField === key.sorter ? props.sortOrder : ''}`} />}
                  </span>
                </th>))}
              {props.handleRemoveItem && <th></th>}
              {props.actions && props.actions.length > 0 && <th></th>}
            </tr>
          </thead>
          <tbody>
            {props.listItems.map(listItem => (
              <tr key={listItem.id} className={listItem.deprecated ? 'deprecated' : ''}>
                {props.keys.map(key => <td className={"table-data"} key={`td-${key.id}`} onClick={() => props.handleRowClick(listItem.id)}>{this.selectData(listItem, key.id)}</td>)}
                {props.actions && props.actions.length > 0 &&
                  props.actions.map(action => {
                    const shouldDeprecate = action.id === identifiers.DEPRECATED && !listItem.deprecated;
                    const shouldUndeprecate = action.id === identifiers.UNDEPRECATED && listItem.deprecated;
                    if (shouldDeprecate || shouldUndeprecate || (action.id !== identifiers.DEPRECATED && action.id !== identifiers.UNDEPRECATED)) {
                      return <td className="remove-list-item table-data" key={action.id} >
                        <Modal
                          id={action.id}
                          modalButtonClassName={action.buttonClass}
                          modalButtonText={action.label}
                          handlePrimaryButton={() => action.handleAction(listItem)}
                          primaryButtonText={action.primaryButtonText}
                          secondaryButtonText={strings.CANCEL}
                          secondaryButtonClassName="btn-light"
                          primaryButtonClassName={action.primaryButtonClassName}
                        >
                          <p>{strings.formatString(action.text, { item: <span className={`text-danger`}>{listItem[props.deleteIdentifier]}</span> })}</p>
                        </Modal>
                      </td>;
                    }
                    return null;
                  })
                }
                {props.handleRemoveItem && <td className="remove-list-item table-data">
                  <Modal
                    id="delete"
                    icon={deleteIcon}
                    modalButtonText=""
                    handlePrimaryButton={() => props.handleRemoveItem(listItem.id, listItem[props.deleteIdentifier])}
                    primaryButtonText={strings.DELETE}
                    secondaryButtonText={strings.CANCEL}
                    secondaryButtonClassName="btn-light"
                    primaryButtonClassName="btn-danger"
                  >
                    <p>{strings.formatString(strings.DELETE_CONFIRMATION, { item: <span className="text-danger">{listItem[props.deleteIdentifier]}</span> })}</p>
                  </Modal>
                </td>}
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
}

Table.propTypes = {
  keys: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    value: PropTypes.string,
    isSortable: PropTypes.bool,
  })).isRequired,
  listItems: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool, PropTypes.object])).isRequired,
  handleRowClick: PropTypes.func.isRequired,
  dateFormat: PropTypes.string,
  handleRemoveItem: PropTypes.func,
  maxTextLength: PropTypes.number,
  deleteIdentifier: PropTypes.string,
  sortOrder: PropTypes.string,
  sortField: PropTypes.string,
};

Table.defaultProps = {
  dateFormat: 'DD/MM/YYYY',
  handleRemoveItem: null,
  maxTextLength: 50,
  deleteIdentifier: '',
  sortOrder: '',
  sortField: '',
};

export default Table;
