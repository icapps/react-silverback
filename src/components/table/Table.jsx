import React from 'react';
import PropTypes from 'prop-types';
import format from 'date-fns/format';
import isDate from 'date-fns/is_date';
import { strings } from '../../utils';
import { Modal } from '../index';
import './table.css';

const checkIcon = require('../../assets/images/check.svg');
const crossIcon = require('../../assets/images/cross.svg');
const deleteIcon = require('../../assets/images/delete.svg');

const SORT_ASC = 'asc';
const SORT_DESC = 'desc';

class Table extends React.Component {
  constructor() {
    super();
    this.state = {
      sortField: '',
      sortOrder: SORT_ASC,
    };
  }

  sort = (sortField, isSortable) => {
    if (isSortable) {
      const sortOrder = (this.state.sortField === sortField) ? (this.state.sortOrder === SORT_ASC ? SORT_DESC : SORT_ASC) : SORT_ASC;
      this.setState({
        sortField,
        sortOrder,
      });
      this.props.handleSort(sortField, sortOrder);
    }
  }

  renderData = data => {
    switch (typeof data) {
      case 'object': return isDate(data) ? format(data, this.props.dateFormat) : data.toString();
      case 'string': return data.length < this.props.maxTextLength ? data : `${data.substring(0, this.props.maxTextLength)}...`;
      case 'boolean': return <img src={data ? checkIcon : crossIcon} alt={`${data}`} />;
      default: return data;
    }
  }
  render() {
    const { props, state } = this;
    return (
      <div className="table-responsive">
        <table className="table table-hover">
          <thead>
            <tr>
              {props.keys.map(key => (
                <th scope="col" key={key.id}>
                  <span className={`key ${key.isSortable ? 'sortable-key' : ''} ${state.sortField === key.id ? 'active-key' : ''}`} onClick={() => this.sort(key.id, key.isSortable)}>
                    {key.value}
                    {key.isSortable && <span className={`sort ${state.sortField === key.id ? state.sortOrder : ''}`} />}
                  </span>
                </th>))}
              {props.handleRemoveItem && <th></th>}
              {props.actions && props.actions.length > 0 && <th></th>}
            </tr>
          </thead>
          <tbody>
            {props.listItems.map(listItem => (
              <tr key={listItem.id} className={listItem.deprecated ? 'deprecated' : ''}>
                {props.keys.map(key => <td className={"table-data"} key={`td-${key.id}`} onClick={() => props.handleRowClick(listItem.id)}>{this.renderData(listItem[key.id])}</td>)}
                {props.actions && props.actions.length > 0 && (listItem.deprecated ? <td /> :
                  <td className="remove-list-item table-data">
                    {props.actions.map(action => <Modal
                      key={action.id}
                      id={action.id}
                      modalButtonClassName={action.buttonClass}
                      modalButtonText={action.label}
                      handlePrimaryButton={() => action.handleAction(listItem.id)}
                      primaryButtonText={action.primaryButtonText}
                      secondaryButtonText={strings.CANCEL}
                      secondaryButtonClassName="btn-light"
                      primaryButtonClassName={action.buttonClass}
                    >
                      <p>{strings.formatString(action.text, { item: <span className="text-danger">{listItem[props.deleteIdentifier]}</span> })}</p>
                    </Modal>)}
                  </td>
                )}
                {props.handleRemoveItem && <td className="remove-list-item table-data">
                  <Modal
                    id="delete"
                    icon={deleteIcon}
                    modalButtonText=""
                    handlePrimaryButton={() => props.handleRemoveItem(listItem.id)}
                    primaryButtonText={strings.DELETE}
                    secondaryButtonText={strings.CANCEL}
                    secondaryButtonClassName="btn-light"
                    primaryButtonClassName="btn-danger"
                  >
                    <p>{strings.formatString(strings.DELETE_CONFIRMATION, { item: <span className="text-danger">{listItem[props.deleteIdentifier]}</span> })}</p>
                  </Modal>
                </td>}
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
}

Table.propTypes = {
  keys: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    value: PropTypes.string,
    isSortable: PropTypes.bool,
  })).isRequired,
  listItems: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool, PropTypes.object])).isRequired,
  handleRowClick: PropTypes.func.isRequired,
  dateFormat: PropTypes.string,
  handleRemoveItem: PropTypes.func,
  maxTextLength: PropTypes.number,
  deleteIdentifier: PropTypes.string,
};

Table.defaultProps = {
  dateFormat: 'DD/MM/YYYY',
  handleRemoveItem: null,
  maxTextLength: 50,
  deleteIdentifier: '',
};

export default Table;
