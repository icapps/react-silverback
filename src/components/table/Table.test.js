import React from 'react';
import { shallow, configure } from 'enzyme';
import Table from './Table';
import Adapter from 'enzyme-adapter-react-16';


configure({ adapter: new Adapter() });

describe('Table Component', () => {
  it('should render a Table component', () => {
    const wrapper = shallow(
      <Table
        keys={['key', 'key1', 'key2', 'key3']}
        listItems={[
          { id: 'id1', key: "test 1", key1: 1, key2: false, key3: new Date() },
          { id: 'id2', key: "test 2", key1: 2, key2: true, key3: new Date() },
          { id: 'id3', key: "test 3", key1: 3, key2: true, key3: new Date() },
          { id: 'id4', key: "test 4", key2: false },
          { id: 'id5', key: "test 5", key1: 4, key2: true, key3: new Date() },
        ]}
        handleRowClick={() => { }}
        handleRemoveItem={() => { }}
        handleSort={() => { }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('should render a Table component with custom Date format', () => {
    const wrapper = shallow(
      <Table
        keys={['key', 'key1', 'key2', 'key3']}
        listItems={
          [
            { id: 'id1', key: "test 1", key1: 1, key2: false, key3: new Date() },
            { id: 'id2', key: "test 2", key1: 2, key2: true, key3: new Date() },
            { id: 'id3', key: "test 3", key1: 3, key2: true, key3: new Date() },
            { id: 'id4', key: "test 4", key2: false },
            { id: 'id5', key: "test 5", key1: 4, key2: true, key3: new Date() },
          ]}
        dateFormat="MM/DD/YYYY"
        handleRowClick={() => { }}
        handleRemoveItem={() => { }}
        handleSort={() => { }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
