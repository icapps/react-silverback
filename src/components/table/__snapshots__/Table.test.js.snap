// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component should render a Table component 1`] = `ShallowWrapper {}`;

exports[`Table Component should render a Table component with custom Date format 1`] = `ShallowWrapper {}`;

exports[`Table Component should render a Table component with custom maximum text length 1`] = `ShallowWrapper {}`;
